### DISCUSSION TITLE
[zfd] How to have a generic function accepting a zod-form-data object?

### DISCUSSION THREAD

**Discussion by CapitaineToinon** [airjp73/rvf#247] [Q&A]
Using zod-form-data@2.0.0, how can I have a function expecting a zod-form-data object that is generic? Here is some code sample of what I'm trying to do. 

Here is some code example: 

```ts
import { z } from 'zod'
import { zfd } from 'zod-form-data'

const schema = zfd.formData({
    email: zfd.text(),
})


function attempt1<T extends z.AnyZodObject>(schema: ReturnType<typeof zfd.formData>): z.infer<typeof schema> {
    // this works but schema is now of type z.ZodEffects<z.ZodTypeAny, any, FormDataLikeInput>
    // I can't pass T into the type here so attempt1 is not generic
    // would love something like ReturnType<typeof zfd.formData<T>> maybe?
    // Or expose FormDataType as generic so I could do schema: FormDataType<T>?
}

const a = attempt1(schema); // works but a is any
// const a: any

// Yikes
type FormDataLikeInput = ReturnType<typeof zfd.formData> extends z.ZodEffects<z.ZodTypeAny, any, infer U> ? U : never

function attempt2<T extends z.AnyZodObject>(schema: z.ZodEffects<T, z.output<T>, FormDataLikeInput>): z.infer<typeof schema> {
    return schema.parse("hey")
}

const b = attempt2(schema) // works and is generic!
// const b: { email: string }
```

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgFCiSyIqbHpY56EkC0m0IrxAhjD+TIBjCADsAzvHFCAFgFMQPOAF5axAHQcoIACJ8eACgRk4puAp7AANgC416mHIAeMAwEoANGTRuyZTACuokIwwGJwfI7gMACMADwAKuYucqLE4ijqAIKiAJ4AWiQA8gBGAFZyIQB8BtLyinYASnIwAVCiCblgcnEwXXIQmPZauvpVbnbI6sCimHJQvf2DcHUWVYgmZgD0W3AwMsAZAO7QANYZJQFSshZwh3CiEEdwy33dmYXEAKKYcyHicSmn063RyuQ8ETyEIAYpw9PwADLAU5yACSojAVyqm1MOzgqLgQh4onw8DAPHEGSSMxgED28j2-Tg8igchWdMiCjAsTuGUe8AA5ql5sAhDi4HiTgErPQrBAAG5s8S4FoHUQCuBWZFs5qtdognpvAZDZB0TRw-SJKrrRS5EpyAD84rxRSgyUg4jZsO08J4BoiGSFohFQnZ+MJEGl9GIdNWDTg3tG-ANVqdaD8Igk8CUqk50RitRuijcAG4JbsTlBznBLtneZDcmQ8ZnJBE7MTG03dgBNbXiMhGhMWxHa9GY+CqXVtDr9RbdZamjQjX3rZyONIZIEkH5-GAArfEA1giEdiEzOZugCq6wdcEvcDswcVUD8gWCoXCee5ACZEsl1+kmRgp8pQVNUhb1DwkzqJ8O6VHuiQQlMkYwOOVowsOPBIiiY5YhMmTnvMc7GisRY8OsxhmHArJ6qIpGQeo5JQJ6BgAETyLkrG+OmwhiK2JQqBEMBRD+EEWG45ZwJW1bEvQ9xBiGACEXYRlmNZ2EgFjWHYkhQDMGpoEAA

Doing this because I'm trying to restrict the zod-form-data object to contain a AnyZodObject and not a ZodTypeAny so I can access the inner shape of the formData. Basically the first signature from this type from zod-form-data:

```ts
type FormDataType = {
    <T extends z.ZodRawShape>(shape: T): ZodEffects<ZodObject<T>, z.output<ZodObject<T>>, FormDataLikeInput>;
    <T extends z.ZodTypeAny>(schema: T): ZodEffects<T, z.output<T>,   >;
};
```

I'm open to other ways I could do things. Thanks!

**Comment by airjp73** [comment]
Hi!

I'm not sure I fully understand what you're trying to do here. Could you give a more concrete description of your use-case?

Whatever the solution ends up being, you most likely will want to leave more up to inference and use fewer explicit types. Without more info, here are a couple examples of how you might wrap a formData schema: [Playground link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgFCiSyIqbHpY56EkC0m0IrxAhjD+TIB6EaLGi4ACQCmAGzDSocAIxlMAVwB2AYxjAImuCB4BraQGVtAC2nGA6sBhXzVngoA8AFTjSAHjGlNYgBnFAA6AC0SACUeAHcXN2kAPgAKYNcFAC44TwBKRDI4YrgoaRh1KENkOjCOKBAAET4edMzpPIBuMjQyMm0DYPgMpMsbYzgAXiNTC2tbHgcnRIVUpDAeYOC46GIc5DChqGBNAHNUgrQ84SESuAA9AH4+8VeRKTkFJQAmNS1dfSGYxmMYLJZWKLEADyACMAFbSXReHz+QIhcKQ2EIpE8TQAT2SaQg8Jy+UKdzKFSqtGIdU4zX4qWJcK6PT6A00QxQJCxU2KB2ZiJgaywEAg+zCmnUIBhigu6FZHK5zNBE2mwLm40WjghPPhQtSRGh8NZ7xKTxeb3EH3kijgAGY-jo9AY4BsoMFpAAxektZF+AJBUIHVU8NLBebGUl5CUnTCKLzJcklJXwepNFp8zTSOJwH0NBmtVkU8qVQwRrVhd2e1LpwvXXqpuAVhZ8mq0ustNZFEoLYCyfa1AL+C4AGh6136g3gvH4fOr3t9jJbxlNt3NzxuVrENq+cAALE6Aa6F16mCtpP7UUGMTF4hfw+1o7HNPGoO4Dpj9UjPITk8UmxXHg21qTtl3aYsU2nUUC0zaZs1zfMM0ZSDikpMtm0jHgqx4D1pFrJceFZXop05GcWjPXAL3nXDPUokALxFGFcIlXCoB4PFDUOGBjjOC5LjXO4niAA)
